{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8441d6ef",
   "metadata": {},
   "source": [
    "# Pipline for Breast Cancer data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1b616107",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports \n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "325ba92a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load csv file into df \n",
    "df = pd.read_csv(\"data/breast-cancer-diagnostic.shuf.lrn.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40d80bd1",
   "metadata": {},
   "source": [
    "## Imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "926be57d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.base import BaseEstimator, TransformerMixin\n",
    "\n",
    "class OutlierImputer(BaseEstimator, TransformerMixin):\n",
    "    \"\"\"\n",
    "    Ersetzt Ausreißer (Z-Score > threshold) durch den Median\n",
    "    der jeweiligen Spalte (berechnet aus Trainingsdaten).\n",
    "    \"\"\"\n",
    "    def __init__(self, threshold=3):\n",
    "        self.threshold = threshold\n",
    "        self.medians_ = None\n",
    "        self.numeric_cols_ = None\n",
    "\n",
    "    def fit(self, X, y=None):\n",
    "        # Nur numerische Spalten\n",
    "        self.numeric_cols_ = X.select_dtypes(include='number').columns\n",
    "        # \n",
    "        self.medians_ = X[self.numeric_cols_].median()\n",
    "        return self\n",
    "\n",
    "    def transform(self, X, y=None):\n",
    "        # Kopie, um Original-Daten nicht zu überschreiben\n",
    "        X_transformed = X.copy()\n",
    "        for col in self.numeric_cols_:\n",
    "            mean = X_transformed[col].mean()\n",
    "            std = X_transformed[col].std()\n",
    "\n",
    "            # Z-Scores\n",
    "            z_scores = (X_transformed[col] - mean) / (std if std != 0 else 1e-9)\n",
    "            outlier_idx = np.where(np.abs(z_scores) > self.threshold)[0]  # int positions\n",
    "\n",
    "            # Outlier -> Median\n",
    "            X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
    "\n",
    "            z_scores = (X_transformed[col] - mean) / (std if std != 0 else 1e-9)\n",
    "            \n",
    "        return X_transformed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "284a0535",
   "metadata": {},
   "source": [
    "## Pipline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d1dfa66a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from imblearn.pipeline import Pipeline\n",
    "from imblearn.over_sampling import SMOTE\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "\n",
    "pipeline = Pipeline([\n",
    "    ('outlier_imputer', OutlierImputer(threshold=3)),\n",
    "    ('smote', SMOTE(random_state=123)),          # Minority Over-Sampling\n",
    "    ('scaler', StandardScaler()),               # StandardScaler\n",
    "    ('pca', PCA(n_components=0.95)),            # PCA mit 95% Varianzerhalt\n",
    "    ('classifier', RandomForestClassifier())    # random frorest\n",
    "])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e87a203",
   "metadata": {},
   "source": [
    "## Cross Validation\n",
    "Vorteil: In jedem Fold wird erst der Outlier-Imputer „gefitet“, dann SMOTE auf den jeweiligen Fold angewendet, anschließend skaliert, PCA berechnet und zuletzt das Modell trainiert. Auf dem Test-Fold werden alle Transformationen mit den Parametern des Trainingsfolds angewandt. Dadurch vermeidest du jegliches Data Leakage."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "778e2c4b",
   "metadata": {},
   "source": [
    "## Optimazation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "54e265fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2025-04-16 13:10:36,601] A new study created in memory with name: no-name-46d98217-a959-4e11-acd4-a81d7d331d6e\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:37,462] Trial 0 finished with value: 0.9340752917197955 and parameters: {'outlier_threshold': 4.477169652736231, 'n_estimators': 127, 'max_depth': 19, 'min_samples_split': 9, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9173255125319805}. Best is trial 0 with value: 0.9340752917197955.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:39,123] Trial 1 finished with value: 0.9343095685244032 and parameters: {'outlier_threshold': 4.403239193128821, 'n_estimators': 294, 'max_depth': 15, 'min_samples_split': 4, 'min_samples_leaf': 1, 'max_features': 'log2', 'pca_n_components': 0.9642673609815939}. Best is trial 1 with value: 0.9343095685244032.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:40,468] Trial 2 finished with value: 0.9413388725537072 and parameters: {'outlier_threshold': 2.2913084351079456, 'n_estimators': 206, 'max_depth': 18, 'min_samples_split': 9, 'min_samples_leaf': 4, 'max_features': 'sqrt', 'pca_n_components': 0.9019024833730107}. Best is trial 2 with value: 0.9413388725537072.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:41,493] Trial 3 finished with value: 0.9267329773527567 and parameters: {'outlier_threshold': 3.4641685112811125, 'n_estimators': 151, 'max_depth': 18, 'min_samples_split': 8, 'min_samples_leaf': 3, 'max_features': 'sqrt', 'pca_n_components': 0.9371998825727139}. Best is trial 2 with value: 0.9413388725537072.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:43,227] Trial 4 finished with value: 0.9415232107380455 and parameters: {'outlier_threshold': 3.097376875939277, 'n_estimators': 266, 'max_depth': 6, 'min_samples_split': 6, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9120015204771429}. Best is trial 4 with value: 0.9415232107380455.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:44,280] Trial 5 finished with value: 0.945051970266805 and parameters: {'outlier_threshold': 2.375406127691573, 'n_estimators': 162, 'max_depth': 20, 'min_samples_split': 6, 'min_samples_leaf': 2, 'max_features': 'sqrt', 'pca_n_components': 0.9409660457411743}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:45,161] Trial 6 finished with value: 0.9342549624697971 and parameters: {'outlier_threshold': 2.8589715356521177, 'n_estimators': 109, 'max_depth': 20, 'min_samples_split': 6, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9762539826249073}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:46,139] Trial 7 finished with value: 0.9343095685244032 and parameters: {'outlier_threshold': 4.016422620781249, 'n_estimators': 145, 'max_depth': 4, 'min_samples_split': 3, 'min_samples_leaf': 1, 'max_features': 'log2', 'pca_n_components': 0.970237182063842}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:46,986] Trial 8 finished with value: 0.9307903914464315 and parameters: {'outlier_threshold': 4.267878261903178, 'n_estimators': 118, 'max_depth': 8, 'min_samples_split': 5, 'min_samples_leaf': 3, 'max_features': 'sqrt', 'pca_n_components': 0.9713135611645027}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:47,947] Trial 9 finished with value: 0.9343095685244032 and parameters: {'outlier_threshold': 3.0842588088896075, 'n_estimators': 154, 'max_depth': 11, 'min_samples_split': 3, 'min_samples_leaf': 4, 'max_features': 'log2', 'pca_n_components': 0.9156703347380905}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:48,582] Trial 10 finished with value: 0.9332192144340491 and parameters: {'outlier_threshold': 2.100846447791649, 'n_estimators': 74, 'max_depth': 14, 'min_samples_split': 7, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'pca_n_components': 0.9444459475412513}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:49,965] Trial 11 finished with value: 0.9338580710729056 and parameters: {'outlier_threshold': 2.5237679411874123, 'n_estimators': 229, 'max_depth': 3, 'min_samples_split': 6, 'min_samples_leaf': 2, 'max_features': 'sqrt', 'pca_n_components': 0.9283447712767301}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:51,624] Trial 12 finished with value: 0.9339867432263022 and parameters: {'outlier_threshold': 3.5130877102733375, 'n_estimators': 275, 'max_depth': 7, 'min_samples_split': 7, 'min_samples_leaf': 2, 'max_features': 'sqrt', 'pca_n_components': 0.9564174747531263}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:53,148] Trial 13 finished with value: 0.9266444288592635 and parameters: {'outlier_threshold': 2.715881289131799, 'n_estimators': 243, 'max_depth': 9, 'min_samples_split': 5, 'min_samples_leaf': 3, 'max_features': 'log2', 'pca_n_components': 0.9884779176610946}. Best is trial 5 with value: 0.945051970266805.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:54,396] Trial 14 finished with value: 0.9452258522258523 and parameters: {'outlier_threshold': 3.3025959088240784, 'n_estimators': 195, 'max_depth': 14, 'min_samples_split': 2, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9099093925024171}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:55,604] Trial 15 finished with value: 0.9379622713919407 and parameters: {'outlier_threshold': 3.715924360538446, 'n_estimators': 188, 'max_depth': 14, 'min_samples_split': 2, 'min_samples_leaf': 3, 'max_features': 'sqrt', 'pca_n_components': 0.9306806091042794}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:56,780] Trial 16 finished with value: 0.9305006811204606 and parameters: {'outlier_threshold': 4.963253145723566, 'n_estimators': 186, 'max_depth': 17, 'min_samples_split': 10, 'min_samples_leaf': 1, 'max_features': 'log2', 'pca_n_components': 0.9531032664029725}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:57,348] Trial 17 finished with value: 0.9224863951424351 and parameters: {'outlier_threshold': 2.0484492937811027, 'n_estimators': 56, 'max_depth': 11, 'min_samples_split': 2, 'min_samples_leaf': 2, 'max_features': 'sqrt', 'pca_n_components': 0.9006263384871984}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:58,594] Trial 18 finished with value: 0.9342991122991122 and parameters: {'outlier_threshold': 3.3098770025135056, 'n_estimators': 207, 'max_depth': 16, 'min_samples_split': 4, 'min_samples_leaf': 3, 'max_features': 'log2', 'pca_n_components': 0.9261311412097998}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:10:59,705] Trial 19 finished with value: 0.9375711687860034 and parameters: {'outlier_threshold': 2.424150203787487, 'n_estimators': 173, 'max_depth': 13, 'min_samples_split': 3, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'pca_n_components': 0.9459896130877334}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:00,430] Trial 20 finished with value: 0.9376998409394 and parameters: {'outlier_threshold': 3.8480396880493686, 'n_estimators': 97, 'max_depth': 20, 'min_samples_split': 4, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9397230130756571}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:01,897] Trial 21 finished with value: 0.9340616204665654 and parameters: {'outlier_threshold': 3.054441486269302, 'n_estimators': 254, 'max_depth': 6, 'min_samples_split': 6, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9140791951303702}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:03,338] Trial 22 finished with value: 0.9378841791237381 and parameters: {'outlier_threshold': 3.107154662751881, 'n_estimators': 217, 'max_depth': 10, 'min_samples_split': 7, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9095766214135502}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:04,913] Trial 23 finished with value: 0.9339321371716961 and parameters: {'outlier_threshold': 2.8378398853272038, 'n_estimators': 270, 'max_depth': 6, 'min_samples_split': 5, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9229633651129998}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:06,543] Trial 24 finished with value: 0.9417711587958832 and parameters: {'outlier_threshold': 3.3267630156277552, 'n_estimators': 295, 'max_depth': 12, 'min_samples_split': 8, 'min_samples_leaf': 3, 'max_features': 'log2', 'pca_n_components': 0.9060174678876816}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:08,177] Trial 25 finished with value: 0.92976452100408 and parameters: {'outlier_threshold': 2.625236472718033, 'n_estimators': 294, 'max_depth': 13, 'min_samples_split': 8, 'min_samples_leaf': 4, 'max_features': 'log2', 'pca_n_components': 0.9091322634334238}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:09,222] Trial 26 finished with value: 0.9264137726533317 and parameters: {'outlier_threshold': 3.6570623452047166, 'n_estimators': 169, 'max_depth': 12, 'min_samples_split': 10, 'min_samples_leaf': 3, 'max_features': 'sqrt', 'pca_n_components': 0.92091331041146}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:10,605] Trial 27 finished with value: 0.9268929555489956 and parameters: {'outlier_threshold': 3.3002690945259427, 'n_estimators': 233, 'max_depth': 16, 'min_samples_split': 8, 'min_samples_leaf': 3, 'max_features': 'log2', 'pca_n_components': 0.9058553493580501}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:11,511] Trial 28 finished with value: 0.9419096459096459 and parameters: {'outlier_threshold': 4.016261919437096, 'n_estimators': 137, 'max_depth': 9, 'min_samples_split': 9, 'min_samples_leaf': 3, 'max_features': 'sqrt', 'pca_n_components': 0.9333938564132552}. Best is trial 14 with value: 0.9452258522258523.\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '901034.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '904745.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "/var/folders/rf/g2w9ss695fj3vt63jtcdpjbr0000gn/T/ipykernel_86248/350205609.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '905609.5' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  X_transformed.iloc[outlier_idx, X_transformed.columns.get_loc(col)] = self.medians_[col]\n",
      "[I 2025-04-16 13:11:12,366] Trial 29 finished with value: 0.9377883894328933 and parameters: {'outlier_threshold': 4.69730432902578, 'n_estimators': 134, 'max_depth': 9, 'min_samples_split': 9, 'min_samples_leaf': 4, 'max_features': 'sqrt', 'pca_n_components': 0.9360446718527798}. Best is trial 14 with value: 0.9452258522258523.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beste Parameter: {'outlier_threshold': 3.3025959088240784, 'n_estimators': 195, 'max_depth': 14, 'min_samples_split': 2, 'min_samples_leaf': 2, 'max_features': 'log2', 'pca_n_components': 0.9099093925024171}\n",
      "Bester Score: 0.9452258522258523\n"
     ]
    }
   ],
   "source": [
    "def objective(trial):\n",
    "    # === Hyperparameter per trial vorschlagen\n",
    "    threshold = trial.suggest_float(\"outlier_threshold\", 2.0, 5.0)  # Range für Outlier\n",
    "    n_estimators = trial.suggest_int(\"n_estimators\", 50, 300)\n",
    "    max_depth = trial.suggest_int(\"max_depth\", 3, 20)\n",
    "    min_samples_split = trial.suggest_int(\"min_samples_split\", 2, 10)\n",
    "    min_samples_leaf = trial.suggest_int(\"min_samples_leaf\", 1, 4)\n",
    "    max_features = trial.suggest_categorical(\"max_features\", [\"sqrt\", \"log2\"])\n",
    "    pca_n_components = trial.suggest_float(\"pca_n_components\", 0.90, 0.99)\n",
    "\n",
    "    # === Pipeline zusammenbauen ===\n",
    "    pipeline = Pipeline([\n",
    "        (\"outlier_imputer\", OutlierImputer(threshold=threshold)),\n",
    "        (\"smote\", SMOTE(random_state=42)),\n",
    "        (\"scaler\", StandardScaler()),\n",
    "        (\"pca\", PCA(n_components=pca_n_components)),\n",
    "        (\"rf\", RandomForestClassifier(\n",
    "            n_estimators=n_estimators,\n",
    "            max_depth=max_depth,\n",
    "            min_samples_split=min_samples_split,\n",
    "            min_samples_leaf=min_samples_leaf,\n",
    "            max_features=max_features,\n",
    "            random_state=42,\n",
    "            n_jobs=-1\n",
    "        ))\n",
    "    ])\n",
    "\n",
    "    # === Cross Validation ===\n",
    "    # Die rohen X und y (nicht vorher skaliert/PCA etc.)\n",
    "    cv_score = cross_val_score(\n",
    "        pipeline, \n",
    "        X, \n",
    "        y, \n",
    "        cv=5,                # z.B. 5-Fold CV\n",
    "        scoring='f1_macro'   # oder f1_weighted, accuracy, etc.\n",
    "    ).mean()\n",
    "\n",
    "    return cv_score\n",
    "\n",
    "# === Optuna Study starten ===\n",
    "study = optuna.create_study(direction=\"maximize\")\n",
    "study.optimize(objective, n_trials=30)\n",
    "\n",
    "# Ergebnis\n",
    "print(\"Beste Parameter:\", study.best_params)\n",
    "print(\"Bester Score:\", study.best_value)\n",
    "\n",
    "# Optional: Visualisierungen\n",
    "# fig1 = optuna.visualization.plot_optimization_history(study)\n",
    "# fig1.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
